solutions/gold/csa-MountainTime.mdx

```diff
@@ -0,0 +1,162 @@
+
+We can break it into two cases.
+
+1. The tallest mountains in $A$ and $B$ are the **same** height. In this case, we'll combine all the cells into one component. In the implementation, we use small-to-large merging to minimize the number of cells we have to update.
+
+2. The tallest mountains in $A$ and $B$ are of **different** heights. Without loss of generality, let's assume that mountain $A$ has the tallest mountain. Then, we know the answer for all the cells in component $B$ because $x$ is the tallest mountain that can connect them both.
+
+To simplify the process, we'll root each component at its maximal cell and store indices of components directly in the DSU. This way, we can easily iterate over cells
+we've found the answer for.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(NM\log{NM})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+// BeginCodeSnip{DSU}
+struct DSU {
+	vector<int> par;
+	vector<vector<int>> comp;
+	DSU(int N) {
+		par = vector<int>(N);
+		// every node is its own parent initially
+		iota(par.begin(), par.end(), 0);
+
+		comp = vector<vector<int>>(N);
+		for (int i = 0; i < N; i++) { comp[i].push_back(i); }
+	}
+
+	// get root (uses path compression)
+	int get(int x) { return par[x] == x ? x : par[x] = get(par[x]); }
+
+	void unite(int x, int y, vector<int> &grid, vector<int> &ans) {
+		int intermediate = grid[x];
+		x = get(x), y = get(y);
+
+		if (x == y) { return; }
+
+		if (grid[x] > grid[y]) swap(x, y);
+
+		if (grid[y] > grid[x]) {
+			// we have the answer for the smaller component
+			for (int &i : comp[x]) {
+				ans[i] = intermediate > grid[i] ? 0 : grid[i] - intermediate;
+			}
+
+			comp[x].clear();
+			par[x] = y;
+		} else if (grid[x] == grid[y]) {
+			// merge the two components (small to large)
+			if ((int)comp[x].size() < (int)comp[y].size()) { swap(x, y); }
+
+			for (int &i : comp[y]) { comp[x].push_back(i); }
+			comp[y].clear();
+			par[y] = x;
+		}
+	}
+};
+// EndCodeSnip
+
+int main() {
+	int n, m;
+	cin >> n >> m;
+
+	vector<int> grid(n * m);
+	vector<int> ans(n * m);
+
+	DSU dsu(n * m);
+
+	vector<pair<int, int>> mountains;
+	for (int i = 0; i < n * m; i++) {
+		cin >> grid[i];
+		ans[i] = grid[i];
+
+		mountains.emplace_back(grid[i], i);
+	}
+
+	sort(mountains.begin(), mountains.end());
+
+	// iterate through all cells from tallest to shortest
+	while (!mountains.empty()) {
+		pair<int, int> cur = mountains.back();
+		mountains.pop_back();
+
+		int idx = cur.second;
+
+		if (idx % m != 0 && grid[idx - 1] >= grid[idx]) {
+			dsu.unite(idx, idx - 1, grid, ans);  // left
+		}
+
+		if (idx >= m && grid[idx - m] >= grid[idx]) {
+			dsu.unite(idx, idx - m, grid, ans);  // up
+		}
+
+		if (idx % m != m - 1 && grid[idx + 1] >= grid[idx]) {
+			dsu.unite(idx, idx + 1, grid, ans);  // right
+		}
+
+		if (idx < (n * m) - m && grid[idx + m] >= grid[idx]) {
+			dsu.unite(idx, idx + m, grid, ans);  // down
+		}
```

Comment:
maybe add some comments to this?

Reply:
it's all explained in the editorial? + already comments for directions

is there any specific part

Reply:
hm
how about
1. remove the whitespace between the if statements
2. on top add a comment "check for mountains higher than the current one on the..."
3. "left side", "on top", etc.

Reply:
sure!